### ISSUE TITLE
CACertificates generation

### ISSUE THREAD

**Issue by abhishekcodework** [Azure/azure-iot-sdk-c#2038]
Can I change the password(and other information such as Organization, Country, etc) in the .cnf files and the certGen.sh and then use certGen.sh to generate root, intermediate and device keys for production?
If not, how do I create my root,intermediate and device keys in the same fashion as the certGen.sh in a way it could b used in production?

**Comment by danewalton** [comment]
Hi @abhishekcodework 
Production level certificates are a more complicated process in several respects, a few of which are generation, maintenance, and storage. 

### Generation
There are several client certificate options that a user can decide to use or not use. This also includes the certificate key length and signing algorithms. Each of this should be evaluated based on the needs and requirements of a user's project. (EX: smaller key for limited device memory or limited CPU power to encrypt/decrypt).

### Maintenance
Another aspect which should be evaluated based on the requirements of the project. What might the update strategy be for the certificates should they be expired or compromised? How many intermediates might you want? etc

### Storage
In general, using supported, native certificate stores are going to be your best friend here (EX; [Windows Certificate Store](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/certificate-stores) and the linux equivalent). There is a balance here between ease of use and security, from easy and unsafe (plain text certificates hard coded into source code) to more complicated but more secure (writing the certificates to a hardware secure module). These must be evaluated by the user or team to understand what risks and technical undertakings they want to weigh.

### TLDR
This is a long answer seemingly unrelated to the question I understand. So the TLDR I would say here is: no. Just changing passwords or common names etc in those files is not enough for us to say it is enough for production.

**Comment by abhishekcodework** [comment]
Thanks a lot, @danewalton this was very helpful. It would be great if you could share any resources to generate certificates for production.

**Comment by danewalton** [comment]
Oh my bad I had some ready but never pasted them. Here are a few:

### Overview of Certificates with IoT Hub
https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-x509ca-overview

### Concepts related to IoT Hub Certificates
https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-x509ca-concept

### Key Vault Certificate Creation
https://docs.microsoft.com/en-us/azure/key-vault/certificates/create-certificate

**Comment by abhishekcodework** [comment]
Thanks @danewalton

**Comment by danewalton** [comment]
You're very welcome. I'll go ahead and close this for now since it seems to be mostly answered. Please let us know if you would like it reopened!
